File: app/src/main/java/org/cerion/tasklist/data/GoogleTasksApi_Impl.java
package org.cerion.tasklist.data;


import android.util.Log;

import org.jetbrains.annotations.NotNull;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.ParseException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import androidx.annotation.Nullable;

public class GoogleTasksApi_Impl extends GoogleApiBase implements GoogleTasksApi {

    private static final String TAG = GoogleTasksApi_Impl.class.getSimpleName();

    GoogleTasksApi_Impl(String authKey) {
        super(authKey, GoogleApi.TASKS_BASE_URL);
    }

    private static final String FIELD_ID = "id";
    private static final String FIELD_TITLE = "title";
    private static final String FIELD_UPDATED = "updated";
    private static final String FIELD_NOTES = "notes";
    private static final String FIELD_DUE = "due";
    private static final String FIELD_DELETED = "deleted";
    private static final String FIELD_STATUS = "status";
    private static final String FIELD_COMPLETED = "completed";

    @Override
    public boolean delete(@NotNull Task task)
    {
        String sURL = getURL("lists/" + task.getListId() + "/tasks/" + task.getId());
        String result = getInetData(sURL,null, DELETE);

        return (result.length() == 0); //Successful delete does not return anything
    }

    @Override
    public Task insert(@NotNull Task task) throws GoogleApiException {
        String listId = task.getListId();
        //String listId = "@default";
        //if(task.getListId() != null && task.getListId().length() > 0)
        //    listId = task.getListId();

        String sURL = getURL("lists/" + listId + "/tasks");
        JSONObject json = new JSONObject();

        try
        {
            json.put(FIELD_TITLE, task.getTitle());
            json.put(FIELD_NOTES, task.getNotes());
            //Only need to set these if non-default value
            if(task.getHasDueDate())
                json.put(FIELD_DUE, mDateFormat.format(task.getDue()));
            if(task.getCompleted())
                json.put(FIELD_STATUS, "completed");

            JSONObject item = getJSON(sURL,json, POST);
            if(item != null && item.has(FIELD_ID) && item.has("selfLink")) {
                String newId = item.getString(FIELD_ID);

                String tokens[] = item.getString("selfLink").split("/");
                for(int i = 0; i < tokens.length - 1; i++) {
                    String token = tokens[i];
                    if(token.contentEquals("lists")) {
                        String newListId = tokens[i+1];
                        return new Task(newListId,newId);
                    }
                }
            }
        }
        catch (JSONException e)
        {
            Log.e(TAG,"exception",e);
        }

        return null;
    }

    @Override
    public boolean update(@NotNull Task task) throws GoogleApiException {
        String sURL = getURL("lists/" + task.getListId() + "/tasks/" + task.getId());
        JSONObject json = new JSONObject();
        boolean bResult = false;

        try
        {
            json.put(FIELD_ID, task.getId());
            json.put(FIELD_TITLE, task.getTitle());
            json.put(FIELD_NOTES, task.getNotes());
            json.put(FIELD_STATUS, task.getCompleted() ? "completed" : "needsAction" );

            if(!task.getCompleted())
                json.put(FIELD_COMPLETED, JSONObject.NULL);
            if(task.getHasDueDate())
                json.put(FIELD_DUE, mDateFormat.format(task.getDue()));
            else
                json.put(FIELD_DUE, JSONObject.NULL);

            JSONObject result = getJSON(sURL,json, PATCH);
            if(result != null && task.getId().contentEquals( result.getString("id") ))
                bResult = true;
        }
        catch (JSONException e)
        {
            Log.e(TAG,"",e);
        }

        return bResult;
    }

    @Override
    public List<Task> list(@NotNull String listId, @Nullable Date dtUpdatedMin) throws GoogleApiException {
        String sURL = getURL("lists/" + listId + "/tasks");
        if(dtUpdatedMin != null)
        {
            sURL += "&updatedMin=" + mDateFormat.format(dtUpdatedMin);
            sURL += "&showDeleted=true";
        }

        JSONObject json = getJSON(sURL);
        ArrayList<Task> result = new ArrayList<>();

        if(json != null)
        {
            try
            {
                JSONArray items = null;
                int count = 0;
                if(json.has("items"))
                {
                    items = json.getJSONArray("items");
                    count = items.length();
                }
                Log.d(TAG, count + " tasks");

                for(int i = 0; i < count; i++)
                {
                    JSONObject item = items.getJSONObject(i);
                    Task task = parseItem(item,listId);
                    if(task != null)
                        result.add(task);
                }
            }
            catch (JSONException e) {
                Log.e(TAG,"exception",e);
            }

        }

        return result;
    }

    private Task parseItem(JSONObject item, String listId) throws JSONException {
        try {
            Task task = new Task(listId, item.getString(FIELD_ID));
            task.setTitle(item.getString(FIELD_TITLE));
            task.setUpdated(mDateFormat.parse(item.getString(FIELD_UPDATED)));

            if(item.has(FIELD_NOTES))
                task.setNotes(item.getString(FIELD_NOTES));
            task.setCompleted(item.getString(FIELD_STATUS).equals("completed"));

            if(item.has(FIELD_DUE)) {
                String sDue = item.getString(FIELD_DUE);
                task.setDue(mDateFormat.parse(item.getString(FIELD_DUE)) );
                System.out.println(sDue + " = " + task.getDue());
            }

            if(item.has(FIELD_DELETED))
                task.setDeleted(item.getBoolean(FIELD_DELETED));

            return task;
        }
        catch (ParseException e) {
            Log.e(TAG,"exception",e);
        }

        return null;
    }
}

File: app/src/main/java/org/cerion/tasklist/data/GoogleTasksApi_Impl.kt
package org.cerion.tasklist.data


import android.util.Log
import androidx.annotation.Nullable
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.text.ParseException
import java.util.*

class GoogleTasksApi_Impl internal constructor(authKey: String) : GoogleApiBase(authKey, GoogleApi.TASKS_BASE_URL), GoogleTasksApi {

    override fun delete(task: Task): Boolean {
        val sURL = getURL("lists/" + task.listId + "/tasks/" + task.id)
        val result = getInetData(sURL, null, GoogleApiBase.DELETE)

        return result.isEmpty() //Successful delete does not return anything
    }

    @Throws(GoogleApiException::class)
    override fun insert(task: Task): Task? {
        val listId = task.listId
        //String listId = "@default";
        //if(task.getListId() != null && task.getListId().length() > 0)
        //    listId = task.getListId();

        val sURL = getURL("lists/$listId/tasks")
        val json = JSONObject()

        try {
            json.put(FIELD_TITLE, task.title)
            json.put(FIELD_NOTES, task.notes)
            //Only need to set these if non-default value
            if (task.hasDueDate)
                json.put(FIELD_DUE, mDateFormat.format(task.due))
            if (task.completed)
                json.put(FIELD_STATUS, "completed")

            val item = getJSON(sURL, json, GoogleApiBase.POST)
            if (item.has(FIELD_ID) && item.has("selfLink")) {
                val newId = item.getString(FIELD_ID)

                val tokens = item.getString("selfLink").split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                for (i in 0 until tokens.size - 1) {
                    val token = tokens[i]
                    if (token.contentEquals("lists")) {
                        val newListId = tokens[i + 1]
                        return Task(newListId, newId)
                    }
                }
            }
        } catch (e: JSONException) {
            Log.e(TAG, "exception", e)
        }

        return null
    }

    @Throws(GoogleApiException::class)
    override fun update(task: Task): Boolean {
        val sURL = getURL("lists/" + task.listId + "/tasks/" + task.id)
        val json = JSONObject()
        var bResult = false

        try {
            json.put(FIELD_ID, task.id)
            json.put(FIELD_TITLE, task.title)
            json.put(FIELD_NOTES, task.notes)
            json.put(FIELD_STATUS, if (task.completed) "completed" else "needsAction")

            if (!task.completed)
                json.put(FIELD_COMPLETED, JSONObject.NULL)
            if (task.hasDueDate)
                json.put(FIELD_DUE, mDateFormat.format(task.due))
            else
                json.put(FIELD_DUE, JSONObject.NULL)

            val result = getJSON(sURL, json, GoogleApiBase.PATCH)
            if (task.id.contentEquals(result.getString("id")))
                bResult = true
        } catch (e: JSONException) {
            Log.e(TAG, "", e)
        }

        return bResult
    }

    @Throws(GoogleApiException::class)
    override fun list(listId: String, @Nullable dtUpdatedMin: Date?): List<Task>? {
        var sURL = getURL("lists/$listId/tasks")
        if (dtUpdatedMin != null) {
            sURL += "&updatedMin=" + mDateFormat.format(dtUpdatedMin)
            sURL += "&showDeleted=true"
        }

        val json = getJSON(sURL)
        val result = ArrayList<Task>()

        try {
            var items: JSONArray? = null
            var count = 0
            if (json.has("items")) {
                items = json.getJSONArray("items")
                count = items!!.length()
            }
            Log.d(TAG, count.toString() + " tasks")

            for (i in 0 until count) {
                val item = items!!.getJSONObject(i)
                val task = parseItem(item, listId)
                if (task != null)
                    result.add(task)
            }
        } catch (e: JSONException) {
            Log.e(TAG, "exception", e)
        }

        return result
    }

    @Throws(JSONException::class)
    private fun parseItem(item: JSONObject, listId: String): Task? {
        try {
            val task = Task(listId, item.getString(FIELD_ID))
            task.title = item.getString(FIELD_TITLE)
            task.updated = mDateFormat.parse(item.getString(FIELD_UPDATED))

            if (item.has(FIELD_NOTES))
                task.notes = item.getString(FIELD_NOTES)
            task.completed = item.getString(FIELD_STATUS) == "completed"

            if (item.has(FIELD_DUE)) {
                val sDue = item.getString(FIELD_DUE)
                task.due = mDateFormat.parse(item.getString(FIELD_DUE))
                println(sDue + " = " + task.due)
            }

            if (item.has(FIELD_DELETED))
                task.deleted = item.getBoolean(FIELD_DELETED)

            return task
        } catch (e: ParseException) {
            Log.e(TAG, "exception", e)
        }

        return null
    }

    companion object {

        private val TAG = GoogleTasksApi_Impl::class.java.simpleName

        private const val FIELD_ID = "id"
        private const val FIELD_TITLE = "title"
        private const val FIELD_UPDATED = "updated"
        private const val FIELD_NOTES = "notes"
        private const val FIELD_DUE = "due"
        private const val FIELD_DELETED = "deleted"
        private const val FIELD_STATUS = "status"
        private const val FIELD_COMPLETED = "completed"
    }
}